core/modules/loader/BufferUdp.h

```diff
@@ -0,0 +1,174 @@
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_BUFFERUDP_H
+#define LSST_QSERV_LOADER_BUFFERUDP_H
+
+// system headers
+#include <arpa/inet.h>
+#include <cstring>
+#include <stdexcept>
+#include <memory>
+#include <sstream>
+#include <string>
+
+// third party headers
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+class MsgElement;
+
+
+/// A buffer for reading and writing. Nothing can be read from the buffer until
+/// something has been written to it.
+/// TODO: rename BufferUdp is not really accurate anymore.
+class BufferUdp {
+public:
+    using Ptr = std::shared_ptr<BufferUdp>;
+
+    /// The absolute largest UDP message we would send.
+    /// Usually, they should be much smaller.
+    static size_t const MAX_MSG_SIZE_UDP = 6000;
+
+    /// These are also being used for TCP messages, which can be much larger.
+    static size_t const MAX_MSG_SIZE_TCP = 10000000;
+
+    /// Create the object with a new _buffer with 'length' bytes.
+    explicit BufferUdp(size_t length = MAX_MSG_SIZE_UDP) : _length(length) {
+        _buffer = new char[length];
+        _ourBuffer = true;
+        _setupBuffer();
+    }
+
+    /// Create a BufferUdp object that uses 'buf' as its buffer, with 'length'
+    /// indicating the number of bytes in the buffer. If 'buf' already
+    /// contains valid data, 'validBytes' must be set to how many bytes of the buffer
+    /// are valid.
+    /// If BufferUdp should take ownership of 'buf', i.e. delete 'buf' when it is done,
+    /// call makeOwnerOfBuffer().
+    BufferUdp(char* buf, size_t length, size_t validBytes) : _buffer(buf), _length(length) {
+        _setupBuffer();
+        advanceWriteCursor(validBytes);
+    }
+
+    BufferUdp(BufferUdp const&) = delete;
+    BufferUdp& operator=(BufferUdp const&) = delete;
+
+    ~BufferUdp() { if (_ourBuffer) { delete[] _buffer; } }
+
+    /// Resets the cursors in the buffer so it is effectively empty.
+    void reset() { _setupBuffer(); }
+
+    /// Return true only if this object owns the buffer.
+    bool releaseOwnership();
+
+    /// Make this object is responsible for deleting _buffer.
+    void makeOwnerOfBuffer() { _ourBuffer = true; }
```

Comment:
I agree, but aspects of asio communication make it very difficult to avoid them. I've tried to wrap the buffers up as much as possible. For the sake of efficiency (as these things can be very large), I've left in the ability to just hand the class a loaded buffer to avoid copying.